File: core/src/main/java/edu/wpi/grip/core/operations/python/PythonScriptFile.java
package edu.wpi.grip.core.operations.python;


import edu.wpi.grip.core.OperationMetaData;
import edu.wpi.grip.core.sockets.InputSocket;
import edu.wpi.grip.core.sockets.OutputSocket;
import edu.wpi.grip.core.sockets.SocketHint;

import com.google.auto.value.AutoValue;

import org.python.core.PyException;
import org.python.core.PyFunction;
import org.python.core.PyObject;
import org.python.core.PyString;
import org.python.core.PySystemState;
import org.python.util.PythonInterpreter;

import java.io.IOException;
import java.net.URL;
import java.util.List;
import java.util.Properties;

/**
 * Converts a string of Python Code or a Python File into something the {@link
 * PythonScriptOperation} can handle.
 */
@AutoValue
public abstract class PythonScriptFile {

  /**
   * Template for custom python operations. Includes imports for sockets, as well as OpenCV
   * core and image processing classes.
   *
   * <p>The sample operation is a simple arithmetic "add" that hopefully shows how the script
   * should be written.</p>
   */
  public static final String TEMPLATE =
      "import edu.wpi.grip.core.sockets.SocketHints.Inputs as inputs\n"
          + "import edu.wpi.grip.core.sockets.SocketHints.Outputs as outputs\n"
          + "import org.bytedeco.javacpp.opencv_core as opencv_core\n"
          + "import org.bytedeco.javacpp.opencv_imgproc as opencv_imgproc\n\n"
          + "name = \"Addition Sample\"\n"
          + "summary = \"The sample python operation to add two numbers\"\n\n"
          + "inputs = [\n"
          + "    inputs.createNumberSpinnerSocketHint(\"a\", 0.0),\n"
          + "    inputs.createNumberSpinnerSocketHint(\"b\", 0.0),\n"
          + "]\n"
          + "outputs = [\n"
          + "    outputs.createNumberSocketHint(\"sum\", 0.0),\n"
          + "]\n\n\n" // two blank lines
          + "def perform(a, b):\n"
          + "    return a + b\n";

  static {
    Properties pythonProperties = new Properties();
    pythonProperties.setProperty("python.import.site", "false");
    PySystemState.initialize(pythonProperties, null);
  }

  /**
   * @param url The URL to get the script file from.
   * @return The constructed PythonScriptFile.
   * @throws IOException If the URL fails to open.
   */
  public static PythonScriptFile create(URL url) throws IOException {
    final String path = url.getPath();
    final String alternativeName = path.substring(1 + Math.max(path.lastIndexOf('/'), path
        .lastIndexOf('\\')));
    final PythonInterpreter interpreter = new PythonInterpreter();
    interpreter.execfile(url.openStream());
    return create(interpreter, alternativeName);
  }

  /**
   * @param code The code to create the file from.
   * @return The constructed PythonScriptFile.
   * @throws PyException if the code has syntax or runtime errors
   */
  public static PythonScriptFile create(String code) {
    final PythonInterpreter interpreter = new PythonInterpreter();
    interpreter.exec(code);
    return create(interpreter, null);
  }

  @SuppressWarnings("unchecked")
  private static PythonScriptFile create(PythonInterpreter interpreter, String alternativeName) {
    final PyString name = interpreter.get("name", PyString.class);
    final PyString summary = interpreter.get("summary", PyString.class);
    final List<SocketHint<PyObject>> inputSocketHints = interpreter.get("inputs", List.class);
    final List<SocketHint<PyObject>> outputSocketHints = interpreter.get("outputs", List.class);
    final PyFunction performFunction = interpreter.get("perform", PyFunction.class);
    return new AutoValue_PythonScriptFile(
        name == null ? alternativeName : name.toString(),
        summary == null ? "" : summary.toString(),
        inputSocketHints,
        outputSocketHints,
        performFunction);
  }

  public abstract String name();

  public abstract String summary();

  public abstract List<SocketHint<PyObject>> inputSocketHints();

  public abstract List<SocketHint<PyObject>> outputSocketHints();

  public abstract PyFunction performFunction();

  /**
   * Converts this file into a {@link PythonScriptOperation}.
   *
   * @param isf Input Socket Factory
   * @param osf Output Socket Factory
   * @return The meta data for a {@link PythonScriptOperation}
   */
  public final OperationMetaData toOperationMetaData(InputSocket.Factory isf,
                                                     OutputSocket.Factory osf) {
    return new OperationMetaData(
        PythonScriptOperation.descriptionFor(this),
        () -> new PythonScriptOperation(isf, osf, this)
    );
  }
}

File: core/src/main/kotlin/edu/wpi/grip/core/operations/PythonScriptFile.kt
package edu.wpi.grip.core.operations


import edu.wpi.grip.core.OperationMetaData
import edu.wpi.grip.core.sockets.InputSocket
import edu.wpi.grip.core.sockets.OutputSocket
import edu.wpi.grip.core.sockets.SocketHint
import org.python.core.PyFunction
import org.python.core.PyString
import org.python.core.PySystemState
import org.python.util.PythonInterpreter
import java.io.IOException
import java.net.URL
import java.util.*

/**
 * Converts a string of Python Code or a Python File into something the [ ] can handle.
 */
data class PythonScriptFile private constructor(
        val name: String,
        val summary: String,
        val inputSocketHints: List<SocketHint<*>>,
        val outputSocketHints: List<SocketHint<*>>,
        val performFunction: PyFunction
) {


    /**
     * Converts this file into a [PythonScriptOperation].

     * @param isf Input Socket Factory
     * *
     * @param osf Output Socket Factory
     * *
     * @return The meta data for a [PythonScriptOperation]
     */
    fun toOperationMetaData(isf: InputSocket.Factory, osf: OutputSocket.Factory): OperationMetaData {
        return OperationMetaData(PythonScriptOperation.descriptionFor(this)) { PythonScriptOperation(isf, osf, this) }
    }

    private companion object {

        init {
            val pythonProperties = Properties()
            pythonProperties.setProperty("python.import.site", "false")
            PySystemState.initialize(pythonProperties, null)
        }

        /**
         * @param url The URL to get the script file from.
         * *
         * @return The constructed PythonScript file.
         * *
         * @throws IOException If the URL fails to open.
         */
        @JvmStatic
        @Throws(IOException::class)
        fun create(url: URL): PythonScriptFile {
            val path = url.path
            val alternativeName = path.subSequence(
                    1 + Math.max(path.lastIndexOf('/'),
                            path.lastIndexOf('\\')), path.length).toString()
            val interpreter = PythonInterpreter()
            interpreter.execfile(url.openStream())
            return create(interpreter, alternativeName)
        }

        /**
         * @param code The code to create the file from.
         * *
         * @return The constructed PythonScript file.
         */
        @JvmStatic
        fun create(code: String): PythonScriptFile {
            val interpreter = PythonInterpreter()
            interpreter.exec(code)
            return create(interpreter, null)
        }

        private fun convertPyToSocketHint(list: List<*>): List<SocketHint<*>> {
            return list.map {
                if (it is SocketHint<*>) {
                    it
                } else {
                    throw ClassCastException(
                            "Return from socket hint method must be type 'SocketHint' " +
                                    "but was $it")

                }
            }
        }

        private fun create(interpreter: PythonInterpreter, alternativeName: String?):
                PythonScriptFile {
            val name = interpreter.get<PyString>("name", PyString::class.java)
            val summary = interpreter.get<PyString>("summary", PyString::class.java)
            val inputSocketHints =
                    convertPyToSocketHint(interpreter.get<List<*>>("inputs", List::class.java))
            val outputSocketHints =
                    convertPyToSocketHint(interpreter.get<List<*>>("outputs", List::class.java))
            val performFunction = interpreter.get<PyFunction>("perform", PyFunction::class.java)
            return PythonScriptFile(
                    if (name == null) alternativeName!! else name.toString(),
                    if (summary == null) "" else summary.toString(),
                    inputSocketHints,
                    outputSocketHints,
                    performFunction)
        }
    }
}
